client/session.go

```diff
@@ -1429,17 +1526,12 @@ func (s *session) streamCollectedBlocksMetadata(
 		}
 
 		received, ok := metadata[m.id.Hash()]
-		if !ok {
+		if !ok || received.submitted {
 			received = &receivedBlocks{
 				results: make([]*blocksMetadata, 0, peersLen),
 			}
 			metadata[m.id.Hash()] = received
 		}
-		if received.submitted {
```

Comment:
You're right, I don't think this event can happen when going through the bootstrap codepath. It can occur when going through the `FetchBlocksFromPeers` method because the interface doesn't mandate that all the blocks per unique ID be grouped into a single `blocksMetadata` equivalent. One of the new unit tests and integration test both capture this. 

for this fix, I think there's 3 choices:
(1) perform the grouping per id in the session.go code
(2) change the `FetchBlocksFromPeers` method interface
(3) move `received.submitted` up

I tried (1) - 2fc0231, didn't like it mainly because it makes testing a lot more cumbersome. Not against doing (2) but I can't think of a good way to capture the grouping dependency in the interface - I don't like mandating it implicitly via docs or by something explicit like taking in map. Both push more onus on the user of the API. So settled on (3) for now. What do you think?